### PR TITLE
Expanded differentiator tests.

### PR DESCRIPTION
**Pull request by MichaelBroughton**
Currently we don't have a great set of tests, making sure that sampled differentiators are working correctly. Added tests here for `sampled_op` behavior.

I think we should try and get this in before #345 so that we can feel a little more confident about those bigger changes coming in.

### CODE REVIEW

File: tensorflow_quantum/python/differentiators/gradient_test.py [hunk 1]
```diff
@@ -220,5 +236,97 @@ def test_analytic_value_with_simple_circuit(self, differentiator, op):
...
+        # Calculate tfq gradient.
+        symbol_values_tensor = tf.convert_to_tensor(symbol_values_array)
+        with tf.GradientTape() as g:
+            g.watch(symbol_values_tensor)
+            expectations = op(circuit, tf.convert_to_tensor(['alpha']),
+                              symbol_values_tensor, psums,
+                              tf.convert_to_tensor([[num_samples]]))
+        grads = g.gradient(expectations, symbol_values_tensor)
+        ground_truth_grads = np.array([[-1.1839752]])
+        self.assertAllClose(ground_truth_grads, grads, rtol=0.2, atol=0.2)
```
Comment (zaqqwerty) [inline tensorflow_quantum/python/differentiators/gradient_test.py:270]: These tolerances seem a bit too large to me, could the `rtol` be removed and the `atol` be tightened up to something like `0.02`?

File: tensorflow_quantum/python/differentiators/gradient_test.py [hunk 2]
```diff
@@ -27,21 +27,37 @@
...
 ]
 
-OPS = [
+SAMPLED_DIFFS = [
+    linear_combination.ForwardDifference(grid_spacing=0.1),
+    linear_combination.CentralDifference(grid_spacing=0.1),
+    parameter_shift.ParameterShift(),
+]
+
+SAMPLED_DIFFS_TOLS = [0.5, 0.5, 0.2]
```
Comment (zaqqwerty) [inline tensorflow_quantum/python/differentiators/gradient_test.py:44]: Here also the tolerances seem a bit loose, could they be made tighter to like `1/10` as large?

File: tensorflow_quantum/python/differentiators/gradient_test.py [hunk 3]
```diff
@@ -220,5 +236,97 @@ def test_analytic_value_with_simple_circuit(self, differentiator, op):
...
+                tf.convert_to_tensor([[num_samples] * n_ops] * n_programs))
+        tfq_grads = g.gradient(expectations, symbol_values_tensor)
+
+        # calculate gradients in cirq using a very simple forward differencing
+        # scheme
+        cirq_grads = _cirq_simple_finite_difference(circuit_batch,
+                                                    resolver_batch,
+                                                    symbol_names, psums)
+
+        self.assertAllClose(cirq_grads, tfq_grads, rtol=tol, atol=tol)
```
Comment (zaqqwerty) [inline tensorflow_quantum/python/differentiators/gradient_test.py:328]: I think `rtol` should be removed, for random circuits gradients could get near zero and then `rtol` might be flagging things unnecessarily

**Review by zaqqwerty — CHANGES_REQUESTED** [review]
Looks good, just wonder about tightening up the tolerances

**Comment by MichaelBroughton** [comment]
Believe it or not, given the number of samples we are taking , this is about as tight as I can make things. We could up the number of samples, but that would blow up the test time pretty substantially. I'm worried if I make things any tighter the test would get flaky.

**Review by zaqqwerty — APPROVED** [review]
Makes sense, LGTM

**Pull request merged**